tools/wake/cli_options.h

```diff
@@ -0,0 +1,206 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <vector>
+
+#include "gopt/gopt-arg.h"
+#include "gopt/gopt.h"
+
+struct CommandLineOptions {
+  bool check;
+  bool verbose;
+  bool debug;
+  bool quiet;
+  bool wait;
+  bool workspace;
+  bool tty;
+  bool fwarning;
+  int profileh;
+  unsigned int input;
+  unsigned int output;
+  bool last_use;
+  bool last_exe;
+  bool lsp;
+  bool failed;
+  bool script;
+  bool metadata;
+  bool version;
+  bool html;
+  bool global;
+  bool help;
+  bool config;
+  bool debugdb;
+  bool parse;
+  bool tcheck;
+  bool dumpssa;
+  bool optim;
+  bool exports;
+  bool timeline;
+  bool clean;
+  bool list_outputs;
+
+  const char *percent_str;
+  const char *jobs_str;
+  const char *memory_str;
+  const char *heapf;
+  const char *profile;
+  const char *init;
+  const char *chdir;
+  const char *in;
+  const char *exec;
+  const char *job;
+  const char *label;
+  char *shebang;
+  const char *tagdag;
+  const char *tag;
+  const char *api;
+  const char *fd1;
+  const char *fd2;
+  const char *fd3;
+  const char *fd4;
+  const char *fd5;
+
+  std::vector<char *> input_files;
+  std::vector<char *> output_files;
+
+  int argc;
+  char **argv;
+
+  CommandLineOptions(int argc_in, char **argv_in) {
+    argv = argv_in;
+    unsigned int max_pairs = argc_in / 2;
+    input_files.reserve(max_pairs);
+    output_files.reserve(max_pairs);
+
+    // clang-format off
+    struct option options[] {
+      {'p', "percent", GOPT_ARGUMENT_REQUIRED | GOPT_ARGUMENT_NO_HYPHEN},
+      {'j', "jobs", GOPT_ARGUMENT_REQUIRED | GOPT_ARGUMENT_NO_HYPHEN},
+      {'m', "memory", GOPT_ARGUMENT_REQUIRED | GOPT_ARGUMENT_NO_HYPHEN},
+      {'c', "check", GOPT_ARGUMENT_FORBIDDEN},
+      {'v', "verbose", GOPT_ARGUMENT_FORBIDDEN | GOPT_REPEATABLE},
+      {'d', "debug", GOPT_ARGUMENT_FORBIDDEN},
+      {'q', "quiet", GOPT_ARGUMENT_FORBIDDEN},
+      {0, "no-wait", GOPT_ARGUMENT_FORBIDDEN},
+      {0, "no-workspace", GOPT_ARGUMENT_FORBIDDEN},
+      {0, "no-tty", GOPT_ARGUMENT_FORBIDDEN},
+      {0, "fatal-warnings", GOPT_ARGUMENT_FORBIDDEN},
+      {0, "heap-factor", GOPT_ARGUMENT_REQUIRED | GOPT_ARGUMENT_NO_HYPHEN},
+      {0, "profile-heap", GOPT_ARGUMENT_FORBIDDEN | GOPT_REPEATABLE},
+      {0, "profile", GOPT_ARGUMENT_REQUIRED},
+      {'C', "chdir", GOPT_ARGUMENT_REQUIRED},
+      {0, "in", GOPT_ARGUMENT_REQUIRED},
+      {'x', "exec", GOPT_ARGUMENT_REQUIRED},
+      {0, "job", GOPT_ARGUMENT_REQUIRED},
+      {'i', "input", GOPT_ARGUMENT_REQUIRED | GOPT_REPEATABLE_VALUE, input_files.data(), max_pairs},
+      {'o', "output", GOPT_ARGUMENT_REQUIRED | GOPT_REPEATABLE_VALUE, output_files.data(), max_pairs},
```

Comment:
Ah yes...the "things you should not be doing with a vector" that I was expecting. If you call ".size()" on this vector it will return zero and if you use ".at()" on it it will throw an out of bounds exception

Reply:
Here's how I'd fix this

1) construct an std::vector of using the constructor without reserving it, setting nullptr as the default value. Just use argc_in as the size and don't bother dividing by 2.
2) give it the data like you do here
3) copy that into an std::vector<std::string> via emplace_back but break when you hit a nullptr

Reply:
1 and 2 are done, 3 wasn't needed as `clo.input` and `clo.output` store the actual number of args found which is later looped over

Reply:
That's still super gross. The size of those vectors should be the number of arguments they contain. Let's clean that up now rather than later. It's trivial to add and is clearly the right way to do this.